enhancements/installer/signal-cluster-deletion.md

```diff
@@ -0,0 +1,130 @@
+---
+title: signal-cluster-deletion
+authors:
+  - "@abutcher"
+reviewers:
+  - dgoodwin
+  - abhinavdahiya
+approvers:
+  - dgoodwin
+  - abhinavdahiya
+creation-date: 2020-03-30
+last-updated: 2020-04-06
+status: provisional
+---
+
+# Signal Cluster Deletion
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Operators can create external resources in clouds and other services that have no opportunity to be cleaned up when a cluster is uninstalled, causing extra work for cluster maintainers to track down and remove these resources. Signalling cluster deletion would allow operators to clean up external resources during uninstallation and would ensure that total cleanup is an automatic process.
+
+## Motivation
+
+### Goals
+
+Signal that cluster deletion has begun through deletion of a new CRD `ClusterAlive` and wait for that object to be deleted during uninstall before continuing with removing cluster resources such as instances, storage and networking.
+
+### Non-Goals
+
+* Attaching finalizers to operator resources based on the `ClusterAlive` object which would facilitate removal of operator resources.
+* Removing resources other than the `ClusterAlive` object during cluster uninstall.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+#### Story 2
+
+### Implementation Details/Notes/Constraints
+
+#### API
+
+Introduce a new CRD `ClusterAlive` from which operator resources could attach a finalizer initiating cascading delete when the `ClusterAlive` resource is deleted.
+
+```yaml
+apiVersion: v1
+kind: ClusterAlive
+metadata:
+  name: cluster
+spec:
+  blockTeardown: true  # block teardown of operator resources
```

Comment:
What is this? What does it do? How does it relate?

Reply:
> Also we should allow the admin to modify the clusteralive in such a way that destroy cannot proceed, just as a safety check.

This was suggested by @abhinavdahiya as a mechanism for admins to block teardown in destroy when set by an admin. I'll add this description to the enhancement.

Reply:
I don't understand the use case or why i'd want what is being described. so hopefully you can lay that out as well. Who is responding to this value in spec? Whatever set the finalizers? Why delete the object if you don't want those things to respond?

Reply:
https://github.com/openshift/enhancements/pull/265#issuecomment-607953675
https://github.com/openshift/enhancements/pull/265#issuecomment-608032530

Reply:
I read those, but I don't think I get it. Is this supposed to prevent someone from 'accidentally' deleting the Alive object and triggering teardown? Are you trying to make it a 2-step process? What's the use case?

Say `blockTeardown: true`
I try to delete the clusterAlive object
We expect all finalizers on the object to pay attention to spec and just do nothing?
Should they remove themselves from the finalizer list?
Should the ClusterAlive object get deleted? Or should it just live forever with a deletionTimestamp but never be deleted?

And why am I doing this?  I'm perfectly happy to understand answers to these questions inside the enhancement instead of here when ready.

Reply:
so currently when deleting the cluster, using destroy cluster we cleanup all the openshift-core-cluster created resources like the `kubernetes.io/cluster/infra-id: owned`

we don't delete any user created resources that don't users don't opt-in into the ownership, and that is also best-effort.

not with cluster alive and `destroy cluster`, there is higher degree of safety concern as now OLM might delete user created resources, like databases etc.

and therefore, i think it's necessary that the admins be allowed to _stop_ _prevent_ any such action to take place.

my initial recommendation was to add this to spec, such that when the user runs `destroy cluster --shutdown` that installer can fail saying that the admin has rejected such action.

also we could also write a admission plugin that would reject any deletes on the object when this field is set.

the operators like OLM probably shouldn't need to actively reconcile this i.e removing finalizers when this being set, but i don't see why that would be bad...

if we can have all operators always add their finalizers, and installer check this or even better has the admission plugin I think that would be very helpful @eparis

Reply:
I've added these details to the enhancement for `blockTeardown`.